text/0000-api-jwt-authentication.md

```diff
@@ -0,0 +1,322 @@
+the algorithm used to sign the token, and a reference to the public key that
+can be used to verify the token signature. In Fabric, this means that `typ`
+must be set to `JWT`, `alg` must be `ES256` for ECDSA-P256 certificates (the
+only type we'll support initially), and `x5c` must be set to the certificate
+or certificate chain of the client in the form described by
+[RFC7517][RFC7517].
+
+The following is an example of a valid header that includes an ECDSA-P256
+certificate:
+
+```json
+{
+  "alg": "ES256",
+  "typ": "JWT",
+  "x5c": [ "MIICPTCCAeSgAwIBAgICBnowCgYIKoZIzj0EAwIwdTELMAkGA1UEBhMCVVMxCTAHBgNVBAgTADEWMBQGA1UEBxMNU2FuIEZyYW5jaXNjbzEbMBkGA1UECRMSR29sZGVuIEdhdGUgQnJpZGdlMQ4wDAYDVQQREwU5NDAxNjEWMBQGA1UEChMNQ29tcGFueSwgSU5DLjAeFw0yMTAxMTIyMDUyMTNaFw0zMTAxMTIyMDUyMTNaMHUxCzAJBgNVBAYTAlVTMQkwBwYDVQQIEwAxFjAUBgNVBAcTDVNhbiBGcmFuY2lzY28xGzAZBgNVBAkTEkdvbGRlbiBHYXRlIEJyaWRnZTEOMAwGA1UEERMFOTQwMTYxFjAUBgNVBAoTDUNvbXBhbnksIElOQy4wWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAASEC/cWPHMJ+EslGw8dzcNINcJWE2H5eMUud1Yv+I+FmbngLabxCoHoEhnoASvyR4LseRfUrCcK7158nF74hoTho2QwYjAOBgNVHQ8BAf8EBAMCB4AwHQYDVR0lBBYwFAYIKwYBBQUHAwIGCCsGAQUFBwMBMA4GA1UdDgQHBAUBAgMEBjAhBgNVHREEGjAYhwR/AAABhxAAAAAAAAAAAAAAAAAAAAABMAoGCCqGSM49BAMCA0cAMEQCIAnkSckDXh/fDEg+qzKavblyc+wW8p9rGRvgac+i7KuZAiBMJxvuT1EOGNIL2ELFzK7iiP8+GKTPxpFqKHzp8gu2zA==" ]
+}
+```
+
+#### Claims
+
+The claims in the `client_assertion` token must include the following fields:
+
+- `aud` - the intended "audience" for the claim and should be the URL of the of the token endpoint
+- `iss` - the `client_id` associated with the certificate
+- `sub` - the `client_id` associated with the certificate
+- `exp` - the token expiration time expressed as the number of seconds since the Unix Epoch
+- `iat` - the time the assertion was created expressed as the number of seconds since the Unix Epoch
+- `jti` - a unique token identifier that helps prevent replay attacks
+
+The `aud` value provided by the client must match the URL of the token
+endpoint. Because of proxying, the expected value(s) may need to be configured
+on the server.
+
+The `iss` and `sub` fields declare the `client_id` that is being asserted.
+Both fields must contain the same value and should be the _MSPID_ that issued
+the certificate. This information allows us to validate the certificate
+against the correct MSP.
+
+The `exp` time in the token should be checked against the local clock. If the
+expiry is in the past, the assertion must not be accepted. If the expiry time
+is more than some threshold in the future, the assertion must not be accepted.
+
+The `iat` time in the token should be checked against the local clock. If the
+token was issued more than some threshold in the past, the assertion must not
+be accepted. The window between `iat` and `exp` should represent a reasonable
+window of time for authentication - typically less than a minute or two.
+
+The `jti` is generated by the client and is intended to act as a nonce to
```

Comment:
can you elaborate on the replay attack? Who is replaying to who?
Also, what if the JTI is not a UUID but it's a timestamp and the peer simply rejects if the timestamp is too far (in absolute value) from its clock?

Reply:
> can you elaborate on the replay attack? Who is replaying to who?

The client authentication token is a signed assertion. Without `jti` the same token can be used multiple times to acquire access tokens.

Additionally, the `jti` is a required filed for open ID connect client authentication and optional in RFC7523. These are the standards that this proposal builds on.

> jti
> REQUIRED. JWT ID. A unique identifier for the token, which can be used to prevent reuse of the token. These tokens MUST only be used once

>  The JWT MAY contain a "jti" (JWT ID) claim that provides a unique identifier for the token.  The authorization server MAY ensure that JWTs are not replayed by maintaining the set of used "jti" values for the length of time for which the JWT would be considered valid based on the applicable "exp" instant.

---

> Also, what if the JTI is not a UUID but it's a timestamp and the peer simply rejects if the timestamp is too far (in absolute value) from its clock?

The `nbf`, `exp`, and `iat` claims control the time window of validity; `exp` is required, `nbf` and `iat` are optional. `iat` and `exp` are referenced in the proposal already and `nbf` is enforced by token validation libraries.

Reply:
> The client authentication token is a signed assertion. Without jti the same token can be used multiple times to acquire access tokens.


But my question is - what scenario are you trying to protect from? Is it a case where a client had access rights, and later on has been revoked? Or that the token was persisted to disk and compromised?

Reply:
My goal is to protect against a compromised authentication token. Even if that weren't the case, the oidc spec requires the `jti` and states that the tokens must be used once.

Just to make sure we're on the same page, the client assertion JWT is used for authentication and not used as the access token...